File: python/evalforge/runner.py
"""
Runner - executes agents against test cases with async support,
timeouts, retries, and progress reporting.
"""

from __future__ import annotations

import asyncio
import inspect
import logging
import time
from typing import Any, Callable, Optional

from .scorer import Scorer
from .test_case import ScoringCriteria, SuiteResult, TestCase, TestResult

logger = logging.getLogger(__name__)

AgentFn = Callable[[Any], Any]  # sync or async callable that takes input -> output


class Runner:
    """
    Runs a list of TestCases against an agent function and returns a SuiteResult.

    Usage::

        runner = Runner(agent=my_agent_fn, suite_name="My Suite")
        result = asyncio.run(runner.run(test_cases))
    """

    def __init__(
        self,
        agent: AgentFn,
        suite_name: str = "eval",
        default_timeout: float = 30.0,
        default_retries: int = 0,
        concurrency: int = 1,
        scorer: Optional[Scorer] = None,
        on_result: Optional[Callable[[TestResult], None]] = None,
    ):
        """
        Args:
            agent: The agent to evaluate. Can be sync or async.
                   Signature: ``agent(input) -> output``.
            suite_name: Name shown in reports.
            default_timeout: Seconds before a test case times out (per attempt).
            default_retries: Number of retries on error (per test case).
            concurrency: Max parallel test cases (default 1 = sequential).
            scorer: Scorer instance. Defaults to ``Scorer()`` (no LLM judge).
            on_result: Optional callback invoked after each TestResult is ready.
        """
        self.agent = agent
        self.suite_name = suite_name
        self.default_timeout = default_timeout
        self.default_retries = default_retries
        self.concurrency = concurrency
        self.scorer = scorer or Scorer()
        self.on_result = on_result

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def run(
        self,
        test_cases: list[TestCase],
        metadata: dict[str, Any] | None = None,
    ) -> SuiteResult:
        """Run all test cases and return a SuiteResult."""
        suite = SuiteResult(
            suite_name=self.suite_name,
            results=[],
            metadata=metadata or {},
            started_at=time.time(),
        )

        semaphore = asyncio.Semaphore(self.concurrency)

        async def _run_one(tc: TestCase) -> TestResult:
            async with semaphore:
                return await self._run_test_case(tc)

        tasks = [_run_one(tc) for tc in test_cases]
        results = await asyncio.gather(*tasks, return_exceptions=False)

        suite.results = list(results)
        suite.finished_at = time.time()
        return suite

    def run_sync(
        self,
        test_cases: list[TestCase],
        metadata: dict[str, Any] | None = None,
    ) -> SuiteResult:
        """Synchronous wrapper around :meth:`run`."""
        return asyncio.run(self.run(test_cases, metadata))

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    async def _run_test_case(self, tc: TestCase) -> TestResult:
        timeout = tc.timeout_seconds if tc.timeout_seconds is not None else self.default_timeout
        max_retries = tc.max_retries if tc.max_retries else self.default_retries
        retries = 0

        # Setup hook
        if tc.setup:
            try:
                result = tc.setup()
                if inspect.isawaitable(result):
                    await result
            except Exception as exc:
                logger.warning("Setup failed for %s: %s", tc.id, exc)

        last_error: Optional[str] = None
        actual_output: Any = None
        latency_ms: Optional[float] = None

        while retries <= max_retries:
            try:
                start = time.perf_counter()
                actual_output = await asyncio.wait_for(
                    self._invoke_agent(tc.input),
                    timeout=timeout,
                )
                latency_ms = (time.perf_counter() - start) * 1000
                last_error = None
                break
            except asyncio.TimeoutError:
                last_error = f"Timed out after {timeout}s"
                retries += 1
                logger.debug("Test %s timed out (attempt %d)", tc.id, retries)
            except Exception as exc:
                last_error = f"{type(exc).__name__}: {exc}"
                retries += 1
                logger.debug("Test %s raised %s (attempt %d)", tc.id, exc, retries)

        # Teardown hook
        if tc.teardown:
            try:
                result = tc.teardown(actual_output)
                if inspect.isawaitable(result):
                    await result
            except Exception as exc:
                logger.warning("Teardown failed for %s: %s", tc.id, exc)

        if last_error:
            test_result = TestResult(
                test_case_id=tc.id,
                passed=False,
                score=0.0,
                actual_output=actual_output,
                error=last_error,
                latency_ms=latency_ms,
                retries=retries - 1,
            )
        else:
            score = await self._score(tc.scoring, tc.expected_output, actual_output)
            passed = score >= tc.scoring.threshold
            test_result = TestResult(
                test_case_id=tc.id,
                passed=passed,
                score=score,
                actual_output=actual_output,
                latency_ms=latency_ms,
                retries=retries,
            )

        if self.on_result:
            self.on_result(test_result)

        return test_result

    async def _invoke_agent(self, input_data: Any) -> Any:
        """Call the agent, supporting both sync and async callables."""
        if inspect.iscoroutinefunction(self.agent):
            return await self.agent(input_data)
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(None, self.agent, input_data)

    async def _score(
        self,
        criteria: ScoringCriteria,
        expected: Any,
        actual: Any,
    ) -> float:
        try:
            return await self.scorer.score(criteria, expected, actual)
        except Exception as exc:
            logger.error("Scoring failed: %s", exc)
            return 0.0

File: typescript/src/runner.ts
/**
 * runner.ts - Executes agents against test cases with timeouts,
 * retries, concurrency control, and progress callbacks.
 */

import { randomUUID } from "crypto";
import { Scorer } from "./scorer.js";
import {
  SuiteResult,
  TestCase,
  TestResult,
} from "./testCase.js";

export type AgentFn = (input: unknown) => unknown | Promise<unknown>;

export interface RunnerOptions {
  /** The agent to evaluate. */
  agent: AgentFn;
  /** Display name for the suite. */
  suiteName?: string;
  /** Per-test timeout in milliseconds. */
  defaultTimeoutMs?: number;
  /** Default number of retries on error. */
  defaultRetries?: number;
  /** Max parallel test cases. */
  concurrency?: number;
  /** Scorer instance. */
  scorer?: Scorer;
  /** Callback after each result. */
  onResult?: (result: TestResult) => void;
}

export class Runner {
  private agent: AgentFn;
  private suiteName: string;
  private defaultTimeoutMs: number;
  private defaultRetries: number;
  private concurrency: number;
  private scorer: Scorer;
  private onResult?: (result: TestResult) => void;

  constructor(opts: RunnerOptions) {
    this.agent = opts.agent;
    this.suiteName = opts.suiteName ?? "eval";
    this.defaultTimeoutMs = opts.defaultTimeoutMs ?? 30_000;
    this.defaultRetries = opts.defaultRetries ?? 0;
    this.concurrency = opts.concurrency ?? 1;
    this.scorer = opts.scorer ?? new Scorer();
    this.onResult = opts.onResult;
  }

  async run(
    testCases: TestCase[],
    metadata: Record<string, unknown> = {}
  ): Promise<SuiteResult> {
    const suite: SuiteResult = {
      suiteName: this.suiteName,
      runId: randomUUID().slice(0, 12),
      startedAt: Date.now(),
      results: [],
      metadata,
    };

    // Process with concurrency control
    const results = await this.runWithConcurrency(testCases);
    suite.results = results;
    suite.finishedAt = Date.now();
    return suite;
  }

  private async runWithConcurrency(testCases: TestCase[]): Promise<TestResult[]> {
    const results: TestResult[] = new Array(testCases.length);
    let index = 0;

    const workers = Array.from({ length: this.concurrency }, async () => {
      while (true) {
        const i = index++;
        if (i >= testCases.length) break;
        results[i] = await this.runTestCase(testCases[i]);
      }
    });

    await Promise.all(workers);
    return results;
  }

  private async runTestCase(tc: TestCase): Promise<TestResult> {
    const timeoutMs = tc.timeoutMs ?? this.defaultTimeoutMs;
    const maxRetries = tc.maxRetries ?? this.defaultRetries;
    let retries = 0;
    let lastError: string | undefined;
    let actualOutput: unknown;
    let latencyMs: number | undefined;

    // Setup hook
    if (tc.setup) {
      try {
        await Promise.resolve(tc.setup());
      } catch (err) {
        console.warn(`[evalforge] setup failed for ${tc.id}:`, err);
      }
    }

    while (retries <= maxRetries) {
      try {
        const start = performance.now();
        actualOutput = await withTimeout(
          Promise.resolve(this.agent(tc.input)),
          timeoutMs
        );
        latencyMs = performance.now() - start;
        lastError = undefined;
        break;
      } catch (err: unknown) {
        lastError = err instanceof Error ? `${err.constructor.name}: ${err.message}` : String(err);
        retries++;
      }
    }

    // Teardown hook
    if (tc.teardown) {
      try {
        await Promise.resolve(tc.teardown(actualOutput));
      } catch (err) {
        console.warn(`[evalforge] teardown failed for ${tc.id}:`, err);
      }
    }

    let result: TestResult;
    if (lastError) {
      result = {
        testCaseId: tc.id,
        passed: false,
        score: 0,
        actualOutput,
        error: lastError,
        latencyMs,
        retries: retries - 1,
        timestamp: Date.now(),
        metadata: {},
      };
    } else {
      const score = await this.scoreResult(tc, actualOutput);
      const threshold = tc.scoring.threshold ?? 1.0;
      result = {
        testCaseId: tc.id,
        passed: score >= threshold,
        score,
        actualOutput,
        latencyMs,
        retries,
        timestamp: Date.now(),
        metadata: {},
      };
    }

    this.onResult?.(result);
    return result;
  }

  private async scoreResult(tc: TestCase, actual: unknown): Promise<number> {
    try {
      return await this.scorer.score(tc.scoring, tc.expectedOutput, actual);
    } catch (err) {
      console.error(`[evalforge] scoring failed for ${tc.id}:`, err);
      return 0;
    }
  }
}

function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error(`Timed out after ${ms}ms`)),
      ms
    );
    promise.then(
      (v) => { clearTimeout(timer); resolve(v); },
      (e) => { clearTimeout(timer); reject(e); }
    );
  });
}
